discojs/discojs-core/src/aggregator/base.ts

```diff
@@ -0,0 +1,239 @@
+   * next aggregation round.
+   */
+  private readonly eventEmitter: EventEmitter
+
+  protected informant?: AsyncInformant<T>
+  /**
+   * The result promise which, on resolve, will contain the current aggregation result.
+   */
+  protected result: Promise<T>
+  /**
+   * The current aggregation round, used for assessing whether a contribution is recent enough
+   * or not.
+   */
+  protected _round = 0
+
+  protected _communicationRound = 0
+
+  constructor (
+    public readonly task: Task,
+    protected _model?: tf.LayersModel,
+    protected readonly roundCutoff = 0,
+    public readonly communicationRounds = 1
+  ) {
+    this.eventEmitter = new EventEmitter()
+    this.contributions = Map()
+    this._nodes = Set()
+
+    this.result = this.makeResult()
+
+    this.eventEmitter.on('aggregation', () => {
+      this.nextRound()
+    })
+  }
+
+  /**
+   * Adds a node's contribution to the aggregator for a given round.
+   * The contribution will be aggregated during the round's aggregation step.
+   * @param nodeId The node's id
+   * @param contribution The node's contribution
+   * @param round For which round the contribution was made
+   */
+  abstract add (nodeId: client.NodeID, contribution: T, round: number, communicationRound?: number): boolean
+
+  /**
+   * Performs the aggregation step over the received node contributions.
+   * Must store the aggregation's result in the aggregator's result promise.
+   */
+  abstract aggregate (): void
+
+  registerObserver (informant: AsyncInformant<T>): void {
+    this.informant = informant
+  }
+
+  /**
+   * Returns whether the given round is recent enough, dependent on the
+   * aggregator's round cutoff.
+   * @param round The round
+   * @returns True if the round is recent enough, false otherwise
+   */
+  isWithinRoundCutoff (round: number): boolean {
+    return this.round - round <= this.roundCutoff
+  }
+
+  /**
+   * Logs useful messages during the various aggregation steps.
+   * @param step The aggregation step
+   * @param from The node which triggered the logging message
+   */
+  log (step: AggregationStep, from?: client.NodeID): void {
+    switch (step) {
+      case AggregationStep.ADD:
+        console.log(`> Adding contribution from node ${from ?? '"unknown"'} for round (${this.communicationRound}, ${this.round})`)
+        return
+      case AggregationStep.UPDATE:
+        if (from === undefined) {
+          return
+        }
+        console.log(`> Updating contribution from node ${from} for round (${this.communicationRound}, ${this.round})`)
+        return
+      case AggregationStep.AGGREGATE:
+        console.log('*'.repeat(80))
+        console.log(`Buffer is full. Aggregating weights for round (${this.communicationRound}, ${this.round})\n`)
+    }
+  }
+
+  setModel (model: tf.LayersModel): void {
+    this._model = model
+  }
+
+  /**
+   * Adds a node's id to the set of "nodes". A node represents an active neighbor
```

Comment:
to the aggregation group

Reply:
in this case, it is rather the set of active nodes, by opposition to the aggregation group defined by the contributing nodes (subset of active nodes)

I'll make sure that the "active nodes" and "aggregation groups" keywords are used consistently throughout the docs